File: contest/LeetCode_Weekly_Contest_203/RabbitLee/Stone_Game_V.py
class Solution:
    def stoneGameV(self, stoneValue: List[int]) -> int:
        n = len(stoneValue)
        f = [[0 for j in range(n)] for i in range(n)]
        s = [0]
        for v in stoneValue:
            s.append(s[-1] + v)
        s = s[1:]
        
        @functools.lru_cache(None)
        def dp(i: int, j: int) -> int:
            if i == j:
                return 0
            ans = 0
            for k in range(i, j):
                # [i..k] [k+1..j]
                left_sum = s[k] - (s[i - 1] if i > 0 else 0)
                right_sum = s[j] - s[k]
                if left_sum > right_sum:
                    ans = max(ans, right_sum + dp(k + 1, j))
                elif left_sum < right_sum:
                    ans = max(ans, left_sum + dp(i, k))
                else:
                    ans = max(ans, left_sum + dp(i, k), right_sum + dp(k + 1, j))
            return ans
        
        return dp(0, n - 1)

File: contest/LeetCode_Weekly_Contest_203/Stone_Game_V.cpp
const int N = 500;
class Solution {
public:
    
    int f[N + 1][N + 1];
    int s[N + 1];
    
    int stoneGameV(vector<int>& stoneValue) {
        int n = stoneValue.size();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                f[i][j] = -1;
            }
        }
        for (int i = 0; i < n; i++) {
            f[i][i] = 0;
        }
        s[0] = stoneValue[0];
        for (int i = 1; i < n; i++) {
            s[i] = s[i - 1] + stoneValue[i];
        }
        return dp(0, n - 1);
    }
    
    int dp(int i, int j) {
        if (f[i][j] >= 0) return f[i][j]; 
        int ans = 0;
        for (int k = i; k < j; k++) {
            int leftSum = s[k] - (i > 0 ? s[i - 1] : 0);
            int rightSum = s[j] - s[k];
            if (leftSum > rightSum) {
                ans = max(ans, rightSum + dp(k + 1, j));
            } else if (leftSum < rightSum) {
                ans = max(ans, leftSum + dp(i, k));
            } else {
                ans = max(max(ans, leftSum + dp(i, k)), rightSum + dp(k + 1, j));
            }
        }
        f[i][j] = ans;
        return ans;
    }
};

